pupil_src/shared_modules/calibration_routines/calibrate_2d.py

```diff
@@ -0,0 +1,264 @@
+        self.degree = degree
+        self.func = [poly.polyval, poly.polyval2d, poly.polyval3d][self.dof - 1]
+        self.params = np.reshape(params, self.dof * [degree + 1])
+
+    def __call__(self, x):
+
+        return self.func(*x, self.params)
+
+
+class PolynomialBinocular:
+    """
+    The average of two monocular polynomials with independent parameters.
+    """
+
+    def __init__(self, params, dof=2, degree=2):
+
+        self.dof = dof
+        self.degree = degree
+
+        self.mapping_eye_0 = PolynomialMonocular(
+            params[: len(params) // 2], self.dof, self.degree
+        )
+        self.mapping_eye_1 = PolynomialMonocular(
+            params[len(params) // 2 :], self.dof, self.degree
+        )
+
+    def __call__(self, x):
+
+        x0 = x[: self.dof]
+        x1 = x[self.dof :]
+
+        return 1 / 2 * (self.mapping_eye_0(x0) + self.mapping_eye_1(x1))
+
+
+class MappingFunction2D:
+    """
+    Maps 2d pupil positions to 2d screen coordinates, using either monocular or binocular polynomials.
+    """
+
+    def __init__(self, params_x, params_y, dof=2, degree=2, binocular=False):
+
+        if binocular:
+            polynomial_type = PolynomialBinocular
+        else:
+            polynomial_type = PolynomialMonocular
+
+        self.map_to_world_cam_x = polynomial_type(params_x, dof, degree)
+        self.map_to_world_cam_y = polynomial_type(params_y, dof, degree)
+
+    def __call__(self, pt):
+
+        return self.map_to_world_cam_x(pt), self.map_to_world_cam_y(pt)
+
+
+def fit_mapping_polynomials(
+    calibration_data,
+    binocular,
+    degree=2,
+    ignored_terms=(),
+    regularization=0.0,
+    loss_scale=-1,
+):
+    """
+    Fit two polynomials, one for each of the last two columns of the provided calibration data.
+
+    :param calibration_data: Array of floats. The last two columns correspond to screen coordinates of the
+    calibration targets. The remaining columns can contain any data that we want to fit on. In the binocular case,
+    the number of columns of the array must be divisble by two.
+    :param binocular: Boolean. Indicates whether the data is binocular or not.
+    :param degree: Int. The degree of the polynomials used.
+    :param ignored_terms: List of tuples. Specifies terms of the polynomials which should be ignored in the fit.
+    Example: (1,2) ignores term of the form x*y**2.
+    :param regularization: Float. Scales a weight decay term in the total calibration cost.
+    :param loss_scale: Float. Outliers are taken care of by a Cauchy loss function. Roughly speaking, squared
+    residuals which are larger than the given loss scale are not influencing the outcome of the fit.
+    :return: Tuple. First entry is a Boolean, indicating whether the fit was successful. Second entry is a tuple
+    containing the parameters and signature of the fitted polynomials (to be used to initialize MappingFunction2D).
+    """
+
+    calibration_data = np.asarray(calibration_data)
```

Comment:
In your documentation you state that `calibration_data` is already an array. Is this conversion needed?